### DISCUSSION TITLE
How to get a typed output in a ResultHandler.handler?

### DISCUSSION THREAD

**Discussion by andremw** [RobinTail/express-zod-api#2218] [Q&A] [answered]
I understand that _the value_ `output` in the `ResultHandler.handler` (a `FlatObject` in this case) is the output of my endpoint handler (that receives an input and returns an output), whose type is defined by the output schema. But being a `FlatObject`, I can't access the properties with proper type safety.

Looking at the code, it seems that I cannot narrow the type of `output` in the `ResultHandler` (FlatObject). Is there any chance we can type that?

My use case is this:
I want to return different status codes depending on the result of my workflow, which might be of type `type Result = { _type: "Success", data: ... } | {_type: "WorkflowError", error: ... }`.
But `defaultEndpointsFactory` doesn't give me a low-level API to do that, so I thought that maybe I could just return from my handler (created with the default factory) and handle the output inside the `ResultHandler`, checking the `_type` and returning the appropriate response for each variant of my discriminated union.

I feel I'm misusing the lib tho, but I've tried to figure out by checking the docs and examples and couldn't spot what I'm missing.
At the same time, I think my idea of narrowing `FlatObject` makes sense, because once I use a `ResultHandler` to actually send the responses, I'm kinda ditching the output defined in the endpoint handler (in favor of positive/negative schemas), am I not?

Thank you! Love the library!

**Comment by RobinTail** [comment, marked as answer]
Hello @andremw ,

the purpose of ResultHandler is responding consistently for every Endpoint produced on EndpointsFactory having that ResultHandler attached to. Therefore, `output` can not have certain type, except the fact that we know that it is an object, because every `output` schema of Endpoint is required to be based on `z.object()`.

However, you can still narrow it down to whatever you want by implementing Typescript conditions like this:

```ts
new ResultHandler({
  handler: ({ output }) => {
    if ("_type" in output && typeof output._type === "string") {
       // inside this block `output` has type { _type: string }
    }
  }
});
```

I must also say, that you should not return errors as a valid output from your Endpoint, like you mentioned:

> {_type: "WorkflowError", error: ... }

this is the job of ResultHandler.

I recommend to throw from Endpoint, so that ResultHandler will receieve the `error` along with an empty `output`.
Consider using `createHttpError()` for that purpose, @andremw 

```ts
factory.build({
  handler: async () => {
    if (something) {
       throw createHttpError(400, "WorkflowError ...");
    }
  }
});

new ResultHandler({
  handler: (output, error) => {
    if (error) {
      // handle error here
      return;
    }
    // handle no error here
  }
})
```

I hope it helps.
Let me know if it's unclear.

**Reply by andremw** [reply]
Yeah that's what I ended up doing. That makes sense, thanks!
